RIOT-OS/RIOT
# ieee802154/radio_hal: detach hal descriptor from driver

### Contribution description
This PR re-arranges the Radio HAL descriptor to be allocated outside of the private descriptor of radios.

In our current scheme, we usually allocate the upper layer descritptor (netdev, hal, etc) inside the private descriptor of the driver.
Therefore we assume it's possible to cast the private descriptor to the interface (and back). However, this approach has some disadvantages:

1. It requires more RAM than allocating the HAL descriptor close to the network stack. The upper layer descriptor requires a pointer to the "context" of the upper layer (e.g the radio HAL context points to the SubMAC if used within GNRC). At the same time, the network stack require to store pointers to the HAL descriptor, because the network stack descriptors and the HAL don't share the same location. If the HAL is allocated closed to the network stack (e.g in the SubMAC), only a pointer to the private descriptor is needed. After all, all `ieee802154_rf_ops` functions simply use the HAL descriptor (`ieee802154_dev_t`) to fetch the private descriptor. The upper layer can always retrieve the network stack descriptor if they were allocated together (e.g `container_of(...)`).
2. The relation of HAL <-> descriptor is not one to one. E.g there are devices that implement 2 radios (at86rf215). Therefore we can simplify a lot the code if we allocate one private descriptor and point it from 2 HAL instances.
  EDIT: Also, there could be cases were several HAL types point to the same device descriptor. An example of this is radios that support crypto acceleration.

From now on, `ieee802154_dev_t`'s `ctx` was renamed to `priv` in order to store the private descriptor of the radios. All Radio HAL functions can access the private descriptor simply following the pointer. Radio events must be boostrapped in order to pass the pointer to the HAL on ISR (this is already done in the current Radio HAL implementations, but it will require to pass the HAL pointer to the args of the ISR in SPI radios).

A follow-up PR with the WIP HAL implementation of at86rf2xx will show this mechanism.



### Testing procedure
Make sure all radio HAL radios still work properly (cc2538, nrf52840). Use e.g `examples/gnrc_networking`.



### Issues/PRs references
Depends on https://github.com/RIOT-OS/RIOT/pull/16533
